### PR TITLE
Initial support for typing.Annotated.

### PR DESCRIPTION
**Pull request by wRAR**
Requires https://github.com/scrapinghub/andi/pull/25

### CODE REVIEW

File: tests/test_injection.py [hunk 1]
```diff
@@ -37,7 +38,13 @@ def __init__(self, crawler):
             self.crawler = crawler
 
         def __call__(self, to_provide):
-            return [cls(content) if content else cls() for cls in classes]
+            result = []
+            for cls in to_provide:
+                obj = cls(content) if content else cls()
+                if metadata := getattr(cls, "__metadata__", None):
+                    obj.__metadata__ = metadata
```
Comment (wRAR) [inline tests/test_injection.py:45]: We need the same change in all real providers if we want to support annotating their provided classes.

File: scrapy_poet/injection.py [hunk 2]
```diff
@@ -261,7 +264,16 @@ def build_instances_from_providers(
...
-            objs_by_type: Dict[Callable, Any] = {type(obj): obj for obj in objs}
+            objs_by_type: Dict[Callable, Any] = {}
+            for obj in objs:
+                cls = type(obj)
+                if sys.version_info >= (3, 9) and (
+                    metadata := getattr(obj, "__metadata__", None)
+                ):
+                    from typing import Annotated
+
+                    cls = Annotated[cls, *metadata]
```
Comment (wRAR) [inline scrapy_poet/injection.py:275]: Ouch, looks like this is invalid syntax in <3.11.

File: tests/test_injection.py (hunk 1 above)
Comment (wRAR) [reply tests/test_injection.py:45]: Another problem with this is when `obj` has `__slots__` and this breaks.

Comment (wRAR) [reply tests/test_injection.py:45]: As we only return instances from the provider's `__call__`, and we need to specify which of them had which annotations, an alternative to this code would be changing the `__call__` interface to e.g. return a tuple.

File: tests/test_injection.py [hunk 3]
```diff
@@ -37,7 +39,13 @@ def __init__(self, crawler):
             self.crawler = crawler
 
         def __call__(self, to_provide):
-            return [cls(content) if content else cls() for cls in classes]
+            result = []
+            for cls in to_provide:
+                obj = cls(content) if content else cls()
+                if metadata := getattr(cls, "__metadata__", None):
+                    obj = AnnotatedResult(obj, metadata)
```
Comment (wRAR) [inline tests/test_injection.py:46]: We need the same change in all real providers if we want to support annotating their provided classes. Also, most of those providers currently don't check `to_provide`, assuming it's the same as the declared provided class(es) so they will need logic improvement. This needs to include support for returning several instances when they are requested.

File: scrapy_poet/api.py [hunk 4]
```diff
@@ -133,3 +134,14 @@ def parse(*args, item: page_or_item_cls, **kwargs):  # type:ignore
 
     setattr(parse, _CALLBACK_FOR_MARKER, True)
     return parse
+
+
+@dataclass
+class AnnotatedResult:
```
Comment (kmike) [inline scrapy_poet/api.py:140]: If we're exposing it in `__init__`, we should probably add a docstring to this class.

**Review by kmike — APPROVED** [review]

**Review by Gallaecio — APPROVED** [review]

**Pull request merged**